packages/react/src/components/NumberInput/NumberInput.tsx

```diff
@@ -110,11 +280,16 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
   });
 
   if (controlledValue !== prevControlledValue) {
-    setValue(controlledValue);
+    // This cast is safe because a component should never change between being a controlled and
+    // uncontrolled component (https://reactjs.org/link/controlled-components).
+    // This condition can only be reached if controlledValue is not equal to prevControlledValue.
+    // controlledValue can only be undefined AND not equal to prevControlledValue if the above
+    // rule is violated (the component changed from controlled to uncontrolled).
+    setValue(controlledValue!);
```

Comment:
I do think an eslint disable line comment will need to be added here. It can be scoped to the individual rule